### PR TITLE
Validate Subject Alternative Name for upstreams (1.10.x)

### PR DESCRIPTION
**Pull request by freddygv**
These changes ensure that the identity of services dialed is
cryptographically verified.

For all upstreams we validate against SPIFFE IDs in the format used by
Consul's service mesh:

spiffe://<trust-domain>/ns/<namespace>/dc/<datacenter>/svc/<service>

### CODE REVIEW

**Review by rboyer — APPROVED** [review]
LGTM pending changelog

**Comment by freddygv** [comment]
go test that failed in CI is passing locally, going to merge

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/410910.

**Pull request merged**
